pkg/sqlutil/sqlutil.go

```diff
@@ -0,0 +1,61 @@
+package sqlutil
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+
+	"github.com/jmoiron/sqlx"
+)
+
+// Queryer is implemented by [*sqlx.DB], [*sqlx.Tx], & [*sqlx.Conn].
+type Queryer interface {
+	sqlx.ExtContext
+	sqlx.PreparerContext
+	GetContext(ctx context.Context, dest interface{}, query string, args ...interface{}) error
+	SelectContext(ctx context.Context, dest interface{}, query string, args ...interface{}) error
+}
+
+type TxOptions struct {
+	sql.TxOptions
+	OnPanic func(recovered any, rollbackErr error)
+}
+
+// Transact is a helper for executing transactions with a domain specific type.
+// A typical use looks like:
+//
+//	func (d *MyD) Transaction(ctx context.Context, fn func(*MyD) error) (err error) {
+//	  return sqlutil.Transact(ctx, d.new, d.q, nil, fn)
+//	}
+func Transact[D any](ctx context.Context, newD func(Queryer) D, q Queryer, opts *TxOptions, fn func(D) error) (err error) {
+	db, ok := q.(interface {
+		// BeginTxx is implemented by *sqlx.DB & *sqlx.Conn, but not *sqlx.Tx.
```

Comment:
Just for my understanding - if this is the case, what use case is there for `*sqlx.Tx` ?

Reply:
You still have transactional guarantees, they just don't nest.